packages/core/src/utils.js

```diff
@@ -1200,7 +1200,18 @@ export function getMatchingOption(formData, options, rootSchema) {
         })),
       };
 
-      let augmentedSchema;
+      // start augmentedSchema as an object to hold only references
+      let {
```

Comment:
I'm not comfortable with this approach, because I think the schema can have additional keys (either in the current spec or in future JSON Schema spec).

What exactly is this code doing, though -- what do you mean by "hold only references"? References can also refer to items within `properties`, too, I believe.

Reply:
yeah I'm not too fond of that either. Let's say we have a rootSchema like this

```json
{
  "components": {
    "a": {"type": "string"}
  },
  "definitions": {
    "a": {"enum": ["1", "2"]}
  },
  ...
}
```

When we have to validate some `schema` then we need to include `components` and `definitions`, i.e. those need to be included when we call [`isValid`](https://github.com/mr-bjerre/react-jsonschema-form/blob/bbf4a9468bf83b785f3fd00bb8f7aa28397112d2/packages/core/src/utils.js#L1239). So basically I'm trying to let the `augmentedSchema` to be an object only holding `components` and `definitions` by removing reserved JSON schema keywords - I'm not sure how to do that properly.

The issue with the latest release state is that `components` and `definitions` wouldn't be included in `augmentedSchema` so if the object being validated would have any `$ref` then `isValid` would return `false`.

Do you have any suggestions?

Reply:
> When we have to validate some `schema` then we need to include `components` and `definitions`,

I think in that case, we could do:

```
let augmentedSchema = {components: rootSchema.components, definitions: rootSchema.definitions};
```

However, I'm still not sure if just including `components` and `definitions` are enough. Does the JSON Schema spec restrict references to items within `components` and `definitions`? I thought $ref's could refer to anywhere in the schema (not necessarily restricted to those keywords), but maybe I'm mistaken.

Reply:
That's exactly the challenge and the reason that I "discarded" properties from the schema instead of handpicking. The logic being; I know what NOT to pick from the schema but I don't know what to pick. 

Alternatively we'd have to find all $ref in the specific schema being validated and make sure to include those from rootSchema

Reply:
Hmm, but could a `$ref` point to an item within `properties`? If so, even the current approach in this PR wouldn't work.

Reply:
I'm not sure if that's allowed. It's definitely bad practice and I don't think it's allowed. 

And yeah I guess that's basically what we do with Object.assign or destructering now. The hard part is retrieving the `defSchema` which is what I basically attempt

Reply:
You can't just combine the root schema with the current subschema since the root schema can have a `type` for instance while the current subschema doesn't. Then the merged would be valid.

Reply:
Actually, it looks like it's not forbidden by the json schema spec. Because of that, I don't know if it's really considered bad practice:

> The "$ref" keyword is an applicator that is used to reference a statically identified schema. Its results are the results of the referenced schema. [CREF5]

> The value of the "$ref" keyword MUST be a string which is a URI-Reference. Resolved against the current URI base, it produces the URI of the schema to apply. This resolution is safe to perform on schema load, as the process of evaluating an instance cannot change how the reference resolves.

https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.8.2.3.1

It's also supported by rjsf ([playground example](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJhZGRyZXNzIjp7fSwic2hpcHBpbmdfYWRkcmVzcyI6eyJzdHJlZXRfYWRkcmVzcyI6IjIyMUIsIEJha2VyIFN0cmVldCIsImNpdHkiOiJMb25kb24iLCJzdGF0ZSI6Ik4vQSJ9LCJ0cmVlIjp7ImNoaWxkcmVuIjpbeyJuYW1lIjoibGVhZiJ9XSwibmFtZSI6InJvb3QifSwiYmlsbGluZ19hZGRyZXNzIjp7InN0cmVldF9hZGRyZXNzIjoiMjEsIEp1bXAgU3RyZWV0IiwiY2l0eSI6IkJhYmVsIiwic3RhdGUiOiJOZXZlcmxhbmQifX0sInNjaGVtYSI6eyJ0eXBlIjoib2JqZWN0IiwicHJvcGVydGllcyI6eyJhZGRyZXNzIjp7InR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7InN0cmVldF9hZGRyZXNzIjp7InR5cGUiOiJzdHJpbmcifSwiY2l0eSI6eyJ0eXBlIjoic3RyaW5nIn0sInN0YXRlIjp7InR5cGUiOiJzdHJpbmcifX0sInJlcXVpcmVkIjpbInN0cmVldF9hZGRyZXNzIiwiY2l0eSIsInN0YXRlIl19LCJzdGF0ZSI6eyJ0aXRsZSI6IlN0YXRlIChyZWZlcmVuY2VkKSIsIiRyZWYiOiIjL3Byb3BlcnRpZXMvYWRkcmVzcy9wcm9wZXJ0aWVzL3N0YXRlIn19fSwidWlTY2hlbWEiOnsidWk6b3JkZXIiOlsiKiJdfSwidGhlbWUiOiJkZWZhdWx0IiwibGl2ZVNldHRpbmdzIjp7InZhbGlkYXRlIjpmYWxzZSwiZGlzYWJsZSI6ZmFsc2UsIm9taXRFeHRyYURhdGEiOmZhbHNlLCJsaXZlT21pdCI6ZmFsc2V9fQ==)).

Reply:
> You can't just combine the root schema with the current subschema since the root schema can have a `type` for instance while the current subschema doesn't. Then the merged would suddenly have potential wrong type

Right, so the solution I have in mind with the [ajv link](https://ajv.js.org/guide/combining-schemas.html#combining-schemas-with-ref) is to not use an augmentedSchema at all, but rather:
- add the root schema using `addSchema()`, set the $id of the schema to to `rootSchema.json`
- rewrite all ref's in the subschema by prepending their values with something like `rootSchema.json`
- call ajv's `validate` function on the subschema

It's kind of messy, but it will work (but maybe there's a better alternative?)

Reply:
Aha I'm catching your drift now :racing_car: 

I'll have a look later. I wonder if this should be done on a higher level then, like initially. Any thoughts on that?

Reply:
Do you want to change `isValid` to receive the `rootSchema` as well? I'm not really sure where the `addSchema()` should happen since it returns an `ajv` instance which is needed for the validation. The issue is that `isValid` uses a global `ajv` instance.

Reply:
> Do you want to change `isValid` to receive the `rootSchema` as well?

Yeah, I think that makes sense. The fact that the `ajv` instance is global is a separate problem (see https://github.com/rjsf-team/react-jsonschema-form/pull/1286/ and https://github.com/rjsf-team/react-jsonschema-form/issues/1433), but work has stalled on it. I think it's too much for this PR to also make ajv also non-global, though.

How about something like this (so we'd keep the global ajv, just reserve a key to be used for root schemas, then add / remove that schema whenever isValid is called):

```js
export function isValid(schema, data, rootSchema) {
  try {
    ajv.addSchema(rootSchema, "__rjsf_rootSchema");
    // transform $ref's in schema to add "__rjsf_rootSchema" as the prefix
    result = ajv.validate(schema, data);
    ajv.removeSchema("__rjsf_rootSchema");
    return result;
  } catch (e) {
    return false;
  }
}

Reply:
I ended up creating a new ajv instance when using rootSchema. Alternatively we'd have to improve the error handling to make sure the `__rjsf_rootSchema` would always be removed (even though validate raises an error). We could easily do that as well of course (maybe just try catch within the existing catch).

I also had to keep the `augmentedSchema` (as in the latest release). Without it I got a `anyOf` test failing, i.e. "should correctly update inputs for anyOf inside array items after being moved down" (both items resolved to `foo` although the second should resolve to `bar` based on the data).

Reply:
Actually, we can't use `createAjvInstance()` -- we need to use the global ajv instance because it includes any additional meta schemas or custom formats: https://github.com/rjsf-team/react-jsonschema-form/blob/64b89213061f1820c55a42d791446bc80ae8d94b/packages/core/src/validate.js#L182-L203

Yeah, makes sense why we need to keep the `augmentedSchema`, given the additional modifications the code does to it for `allOf` / `required` properties.